plugins/modules/ec2_vpc_vgw.py

```diff
@@ -198,39 +205,54 @@ def get_vgw_info(vgws):
     polling_increment_secs = 15
     max_retries = module.params.get("wait_timeout") // polling_increment_secs
-    status_achieved = False
+    try:
+        wait_for_resource_state(client, module, "vpn_gateway_exists", VpnGatewayIds=vpn_gateway_id)
+        if desired_status == "attached":
+            wait_for_resource_state(
+                client,
+                module,
+                "vpn_gateway_attached",
+                VpnGatewayIds=vpn_gateway_id,
+                delay=polling_increment_secs,
+                max_attempts=max_retries,
+            )
+        elif desired_status == "detached":
+            wait_for_resource_state(
+                client,
+                module,
+                "vpn_gateway_detached",
+                VpnGatewayIds=vpn_gateway_id,
+                delay=polling_increment_secs,
+                max_attempts=max_retries,
+            )
+        else:
+            module.fail_json(msg=f"Unsupported status: {desired_status}")
 
-    for x in range(0, max_retries):
-        try:
-            response = find_vgw(client, module, vpn_gateway_id)
-            if response[0]["VpcAttachments"][0]["State"] == status:
-                status_achieved = True
-                break
-            else:
-                time.sleep(polling_increment_secs)
-        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
-            module.fail_json_aws(e, msg="Failure while waiting for status update")
+        response = find_vgw(client, module, vpn_gateway_id)
+        status_achieved = response[0]["VpcAttachments"][0]["State"] == desired_status
 
-    result = response
-    return status_achieved, result
+    except AnsibleEC2Error as e:
+        module.fail_json_aws(e)
+
+    return status_achieved, response
 
 
 def attach_vgw(client, module, vpn_gateway_id):
     params = dict()
-    params["VpcId"] = module.params.get("vpc_id")
+    vpc_id = module.params.get("vpc_id")
+    if vpc_id:
+        params["VpcId"] = vpc_id
+    if vpn_gateway_id:
+        params["VpnGatewayId"] = vpn_gateway_id
 
     try:
-        # Immediately after a detachment, the EC2 API sometimes will report the VpnGateways[0].State
-        # as available several seconds before actually permitting a new attachment.
-        # So we catch and retry that error.  See https://github.com/ansible/ansible/issues/53185
-        response = VGWRetry.jittered_backoff(retries=5, catch_extra_error_codes=["InvalidParameterValue"])(
-            client.attach_vpn_gateway
-        )(VpnGatewayId=vpn_gateway_id, VpcId=params["VpcId"])
-    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
-        module.fail_json_aws(e, msg="Failed to attach VPC")
+        # response = attach_vpc_vpn_gateway(client, vpn_gateway_id, params["VpcId"])
+        response = attach_vpc_vpn_gateway(client, **params)
+    except AnsibleEC2Error as e:
+        module.fail_json_aws(e)
```

Comment:
```suggestion
        module.fail_json_aws_error(e)
```

Reply:
Shouldn't these errors be handled in the amazon.aws.ec2 utility functions?